### PR TITLE
gnrc_tcp: rewrite API internal messaging

### PR DESCRIPTION
**Pull request by brummer-simon**
Hi,

This PR enables the internal fsm of gnrc_tcp to store messages in mboxes allocated outside of a single tcb. This change allows to use a single mbox to gather events of multiple tcbs. 

This is a preparation for a "soon to follow" API change required to integrate
GNRC_TCP into SOCK. I have tested this PR on native and on the nucleo-f401re. 

This PR does not introduce new features, testing by executing the test scripts in "tests/gnrc_tcp" should be sufficient.

Cheers Simon

### CODE REVIEW

**Comment by miri64** [comment]
> This change allows to use a single mbox to gather events of multiple tcbs.

How and why is this required for `sock_tcp`?

**Comment by brummer-simon** [comment]
It is not required for sock_tcp directly, but for an api change in gnrc_tcp. I want to implement the passive connection opening as tcb queue based listen/accept scheme (like sock_tcp). In need this for the accept implementation.

In case I have multiple TCBs waiting for an incoming SYN, I want to use a single mbox to gather messages sent on connection establishment. That way I can have a single mbox used to determine the affected TCB after a connection was established.
Without this PR I would have to iterate over the TCBs, wait for on each TCB on incomings events with a timeout. If the timeout expires I'll have to wait on the with the next TCB, otherwise I'll have accept the connection.

All in all I think this PR allows a much simpler implementation of accept and less delay for accepting incoming connections.

**Comment by miri64** [comment]
Thank you for the explanation. This makes the direction of this PR much clearer.

File: sys/net/gnrc/transport_layer/tcp/internal/common.h [hunk 1]
```diff
@@ -39,14 +39,18 @@ extern "C" {
  */
 #define PORT_UNSPEC (0)
 
+/**
+ * @brief Message queue size for messaging between thread contexts
+ */
+#define TCP_MSG_QUEUE_SIZE (8U)
```
Comment (miri64) [inline sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: Why did you remove the configurability of the message size(s) I would rather split it up into the different modes the `mbox` is used (open, send, recv if I see correctly) and provide a config value for each. Also, since they are stored on stack now, we should rethink if a default of 8 is not maybe to large.

Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: I move the define into the complation unit gnrc_tcp.c because this is the only location were the define is in use. 
I'll agree with you that splitting the defines is a good Idea.

Comment (miri64) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: For me this is less about usage as about configurability. Would be nice to still have these queue size configurable as they _might_ have an impact on performance.

Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: In terms of configurability:
Should i move it to the public api "sys/include/net/gnrc/transport_layer/tcp/config.h" and add kconfig support then?

Comment (miri64) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: Yes please, otherwise people won't find it ;-).

Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: Okay, I'll move the define and  setup it to four. I will not introduct a second define for a larger mbox in this PR.

The second define will be added then adding gnrc_tcp_accept, this is the only function thats requires a larger mbox.

**Comment by miri64** [comment]
Here as well: please provide a more human readable commit summary and PR title.

File: sys/net/gnrc/transport_layer/tcp/internal/common.h (hunk 1 above)
Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: While I am at it:  What about making the eventloop message queue configurable?

Comment (miri64) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: It isn't yet :eyes:?

Comment (brummer-simon) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: I am pretty sure it is not configureable externally.

Comment (miri64) [reply sys/net/gnrc/transport_layer/tcp/internal/common.h:45]: Then it should be :-)

**Comment by brummer-simon** [comment]
I changed the name and addressed the discussed issues. Default message queue size is now 4 and kconfig support is added.

**Comment by miri64** [comment]
@leandrolanzieri do you want to have a quick looks regarding the Kconfig integration?

**Comment by leandrolanzieri** [comment]
> @leandrolanzieri do you want to have a quick looks regarding the Kconfig integration?

LGTM!

**Comment by miri64** [comment]
Then let's go.
